x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/topn/TopNOperator.java

```diff
@@ -349,12 +357,22 @@ public void addInput(Page page) {
 
             for (int i = 0; i < page.getPositionCount(); i++) {
                 if (spare == null) {
-                    spare = new Row(breaker, sortOrders);
+                    spare = new Row(breaker, sortOrders, spareKeysPreAllocSize, spareValuesPreAllocSize);
                 } else {
                     spare.keys.clear();
                     spare.values.clear();
+                    assert spare.keys.clearResizes();
+                    assert spare.values.clearResizes();
                 }
                 rowFiller.row(i, spare);
+                assert addRowResizes(spare.values.getResizes() + spare.keys.getResizes());
+
+                // When rows are very long, appending the values one by one can lead to lots of allocations.
+                // To avoid this, pre-allocate at least as much size as in the last seen row.
+                // Let the pre-allocation size decay in case we only have 1 huge row and smaller rows otherwise.
+                spareKeysPreAllocSize = Math.max(spare.keys.length(), spareKeysPreAllocSize / 2);
```

Comment:
Nice!

Reply:
I think this is a better compromise than I was thinking. I like the idea of a buffer, but we don't have time for that at the moment.